src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java

```diff
@@ -140,6 +150,48 @@ public static void loadNative() {
     private static final Map<String,PKCS11> moduleMap =
         new HashMap<String,PKCS11>();
 
+    static boolean isKey(CK_ATTRIBUTE[] attrs) {
+        for (CK_ATTRIBUTE attr : attrs) {
+            if ((attr.type == CKA_CLASS) && (attr.getLong() == CKO_SECRET_KEY)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // This is the SunPKCS11 provider instance
+    // there can only be a single PKCS11 provider in
+    // FIPS mode.
+    private static SunPKCS11 mysunpkcs11;
+
+    private static final class InnerPKCS11 extends PKCS11 implements Consumer<SunPKCS11> {
+        InnerPKCS11(String pkcs11ModulePath, String functionListName) throws IOException {
+            super(pkcs11ModulePath, functionListName);
+        }
+
+        // Set PKCS11 instance to FIPS mode, called by SunPKCS11 provider.
+        @Override
+        public void accept(SunPKCS11 sunpkcs11) {
+            mysunpkcs11 = sunpkcs11;
+        }
+
+        // Overriding the JNI method C_CreateObject so that first check if FIPS mode is on and the object is a
+        // secret key, in which case invoke the importKey method in SunPKCS11 provider to import the secret key
+        // into the PKCS11 device.
+        public synchronized long C_CreateObject(long hSession, CK_ATTRIBUTE[] pTemplate) throws PKCS11Exception {
+            if ((mysunpkcs11 != null) && isKey(pTemplate)) {
+                try {
+                    Method method = mysunpkcs11.getClass().getDeclaredMethod("importKey", long.class, CK_ATTRIBUTE[].class);
```

Comment:
Why isn't `importKey()` public so reflection is not needed (the package is not exported from this module)?

Reply:
@keithc-ca 

This is a port PR from Alon previous JDK11 PR https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/542. And in Alon’s JDK11 PR, he defined the method “importKey” as the default access modifier, so it can not be accessed out of the packages directly.

My understanding for why this method is not defined as public is, this method “importKey” and the other method “exportKey” are only used for Semeru in FIPS mode, for our internal user such as Liberty, whose Lightweight Third Party Authentication (LTPA) feature requires importing and exporting cryptographic keys in Semeru FIPS mode. Right now, they may not be the kind of methods which we can public and record in the Java document.

Reply:
Yes, if the method is `public` it can be accessed from other packages, but because the defining package is not exported, it cannot (normally) be accessed from other modules.